### PR TITLE
[SPARK-49281][SQL] Optimze parquet binary getBytes with getBytesUnsafe to avoid copy cost

### PR DESCRIPTION
**Pull request by xuzifu666**
### What changes were proposed in this pull request?

Optimze parquet binary getBytes with getBytesUnsafe to avoid copy cost 

### Why are the changes needed?

Binary##getBytes()  need copy all the internal bytes and getBytesUnsafe not need with more quick performence.
```
 /**
   * Variant of getBytes() that avoids copying backing data structure by returning
   * backing byte[] of the Binary. Do not modify backing byte[] unless you know what
   * you are doing.
   *
   * @return backing byte[] of correct size, with an offset of 0, if possible, else returns result of getBytes()
   */
  public abstract byte[] getBytesUnsafe();
```

Add a benchmark test for it:
before the pr:
```

Java HotSpot(TM) 64-Bit Server VM 17.0.12+8-LTS-286 on Mac OS X 13.6.3
Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
Parquet dictionary:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Read binary dictionary                             9137           9706         401         11.5          88.2      1.0X
```

after the pr:
```

Java HotSpot(TM) 64-Bit Server VM 17.0.12+8-LTS-286 on Mac OS X 13.6.3
Intel(R) Core(TM) i7-1068NG7 CPU @ 2.30GHz
Parquet dictionary:                       Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
Read binary dictionary                             9042           9502         370         10.9          87.1       1.0X
```

we found that after the pr speed had a improvemence.

image

This optizme also can refer from other patches which can improve parquet read pefermance, such as: https://github.com/apache/paimon/pull/3509

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Other test cover this pr.

### Was this patch authored or co-authored using generative AI tooling?

No

### CODE REVIEW

**Comment by xuzifu666** [comment]
@yaooqinn PTAL

**Comment by wangyum** [comment]
@xuzifu666 Do you have some benchmark numbers?

**Comment by xuzifu666** [comment]
> @xuzifu666 Do you have some benchmark numbers?

@wangyum Had add benchmark test for it in description.

**Comment by xuzifu666** [comment]
cc @ulysses-you

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala [hunk 1]
```diff
@@ -806,8 +806,8 @@ class ParquetFilters(
               }
 
               override def keep(value: Binary): Boolean = {
-                value != null && UTF8String.fromBytes(value.getBytes).startsWith(
-                  UTF8String.fromBytes(strToBinary.getBytes))
+                value != null && UTF8String.fromBytes(value.getBytesUnsafe).startsWith(
+                  UTF8String.fromBytes(strToBinary.getBytesUnsafe))
```
Comment (wangyum) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/parquet/ParquetFilters.scala:810]: It's `ByteBufferBackedBinary`, and `isBackingBytesReused` is false. This change seems to be an improvement:
https://github.com/apache/parquet-java/blob/master/parquet-column/src/main/java/org/apache/parquet/io/api/Binary.java#L473-L493

**Review by cloud-fan — APPROVED** [review]

**Review by wangyum — APPROVED** [review]

**Comment by ulysses-you** [comment]
thanks, merging to master

**Comment by ulysses-you** [comment]
@xuzifu666 the JIRA ID is wrong.. it seems [SPARK-49281](https://issues.apache.org/jira/browse/SPARK-49281) is your ticket..

**Comment by xuzifu666** [comment]
> @xuzifu666 the JIRA ID is wrong.. it seems [SPARK-49281](https://issues.apache.org/jira/browse/SPARK-49281) is your ticket..

OK，Thanks for your fix. @ulysses-you

**Pull request closed without merging**